bugsnag-android-core/src/main/java/com/bugsnag/android/Journal.java

```diff
@@ -0,0 +1,170 @@
+package com.bugsnag.android;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.dslplatform.json.DslJson;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A journal represents a series of commands that will manipulate a document.
+ * Sometimes it's more efficient to store a base document and a list of modifications
+ * to me made to it rather than re-serialize the entire document after every modification.
+ */
+public class Journal {
+    @NonNull
+    public final List<Command> commands;
+
+    public Journal() {
+        commands = new LinkedList<>();
+    }
+
+    /**
+     * Apply this journal to a document.
+     *
+     * This method goes through the journal entries one-by-one, applying their modification
+     * commands to the document. This is a destructive operation that will modify the document.
+     *
+     * Note: If a journal command replaces the root object, the resulting document will be a
+     * different object from the one passed in.
+     *
+     * @param document The document to apply to.
+     * @return The finished document (may be a different object from the one passed in).
+     */
+    public @NonNull Object applyTo(@NonNull Object document) {
+        for (Command command : commands) {
+            document = command.apply(document);
+        }
+        return document;
+    }
+
+    /**
+     * Add a journal command to this journal
+     *
+     * @param command The journal command
+     */
+    public void add(@NonNull Command command) {
+        this.commands.add(command);
+    }
+
+    /**
+     * Clear the journal, removing all journal entries.
+     */
+    public void clear() {
+        this.commands.clear();
+    }
+
+    /**
+     * Serialize the journal to a stream.
+     *
+     * @param out The stream to serialize to.
+     * @throws IOException if an IO exception occurs.
+     */
+    public void serialize(@NonNull OutputStream out) throws IOException {
+        for (Command command : commands) {
+            command.serialize(out);
+            out.write(0);
+            out.flush();
+        }
+    }
+
+    /**
+     * Deserialize from a byte array.
+     *
+     * @param bytes The bytes to deserialize from
+     * @return The journal
+     */
+    public static @NonNull Journal deserialize(@NonNull byte[] bytes, @NonNull Logger logger) {
+        Journal journal = new Journal();
+        int currentStart = 0;
+        for (int i = 0; i < bytes.length; i++) {
+            if (bytes[i] == 0) {
+                try {
+                    byte[] document = Arrays.copyOfRange(bytes, currentStart, i);
+                    journal.add(Command.deserialize(document));
+                    currentStart = i + 1;
+                } catch (Exception exception) {
```

Comment:
What operations specifically can fail here, and what would the ramifications be if it wasn't possible to deserialize a journal entry?

Reply:
Bad JSON from a partial write will be the most common cause. If deserialization of a journal entry fails, we toss this one out and continue on. You end up with partial data, but it's better than no data.